plugins/microservices/msi_tests/libmsi_test_scoped_permission.cpp

```diff
@@ -0,0 +1,274 @@
+
+            // Change permissions to read.
+            {
+                const auto temporary_read_perm = sp{*comm, path, fs::perms::read};
+                IRODS_MSI_ASSERT(fs::perms::read == fs::server::status(*comm, path).permissions().at(0).prms)
+
+                // The user should be able to open the object for read.
+                {
+                    io::server::native_transport tp{*comm};
+                    IRODS_MSI_ASSERT((io::idstream{tp, path}))
+                }
+
+                // The user should no longer be able to open the object for write.
+                {
+                    io::server::native_transport tp{*comm};
+                    IRODS_MSI_ASSERT(!(io::odstream{tp, path, std::ios_base::app}))
+                }
+
+                // Change permissions to null.
+                {
+                    const auto temporary_null_perm = sp{*comm, path, fs::perms::null};
+
+                    // The user should no longer be able to open the object for read.
+                    {
+                        io::server::native_transport tp{*comm};
+                        IRODS_MSI_ASSERT(!(io::idstream{tp, path}))
+                    }
+                }
+
+                // Read permission is restored.
+                IRODS_MSI_ASSERT(fs::perms::read == fs::server::status(*comm, path).permissions().at(0).prms)
+
+                // The user should be able to open the object for read again.
+                {
+                    io::server::native_transport tp{*comm};
+                    IRODS_MSI_ASSERT((io::idstream{tp, path}))
+                }
+
+                // The user should still be unable to open the object for write.
+                {
+                    io::server::native_transport tp{*comm};
+                    IRODS_MSI_ASSERT(!(io::odstream{tp, path, std::ios_base::app}))
+                }
+            }
+
+            // Write permission is restored.
+            IRODS_MSI_ASSERT(fs::perms::write == fs::server::status(*comm, path).permissions().at(0).prms)
+
+            // The user should be able to open the object for write again.
+            {
+                io::server::native_transport tp{*comm};
+                IRODS_MSI_ASSERT((io::odstream{tp, path, std::ios_base::app}))
+            }
+
+            // The user should still be unable to remove the object.
+            try {
+                fs::server::remove(*comm, path);
+            }
+            catch (const fs::filesystem_error& e) {
+            }
+
+            IRODS_MSI_ASSERT(fs::server::exists(*comm, path))
+        }
+
+        // Own permission is restored.
+        IRODS_MSI_ASSERT(fs::perms::own == fs::server::status(*comm, path).permissions().at(0).prms)
+
+        IRODS_MSI_TEST_END
+    } // test_basic_usage
+
+    auto msi_impl(RuleExecInfo* _rei) -> int
+    {
+        IRODS_MSI_TEST_CASE([](RuleExecInfo* _rei) { return !(_rei && _rei->rsComm); }, _rei)
```

Comment:
What does this produce in the log file? Is it easy to recognize?

Reply:
Oh, I guess I need to wrap this with something in order to see anything in the logs. Currently, nothing is being printed out because it's not using the other macros. Should I just get rid of it? It's a sanity check in a test, so, this assertion being false would end up crashing the agent later due to nullptr dereferences. But that would indicate an issue in the microservice plugin interface or the REP which called it.

Reply:
I think it's fine to remove. If a nullptr is passed, then we have bigger problems. Also, I think it's safe to assume it is always valid.

Thoughts?

Reply:
Yep, makes sense to me.

Reply:
Done

Reply:
maybe make a note in the docs to make this assumption explicit

Reply:
Let's capture that in an issue in irods_docs. That info is likely for the developer docs.

Reply:
Done. https://github.com/irods/irods_docs/issues/205